### PR TITLE
Added ability to search objects using regular expressions

### PR DESCRIPTION
**Pull request by sm15455**
#708

### CODE REVIEW

File: src/GUI/EFCorePowerTools/ViewModels/ObjectTreeViewModel.cs [hunk 1]
```diff
@@ -41,11 +42,20 @@ public class ObjectTreeViewModel : ViewModelBase, IObjectTreeViewModel
...
             foreach (var obj in _objects)
             {
-                obj.IsVisible = string.IsNullOrWhiteSpace(searchText) || obj.DisplayName.ToUpper().Contains(searchText.ToUpper());
+                if (searchMode == SearchMode.Text)
+                {
+                    obj.IsVisible = string.IsNullOrWhiteSpace(searchText) || obj.DisplayName.ToUpper().Contains(searchText.ToUpper());
+                }
+                else
+                {
+                    obj.IsVisible = regex.IsMatch(obj.DisplayName);
```
Comment (ErikEJ) [inline src/GUI/EFCorePowerTools/ViewModels/ObjectTreeViewModel.cs:57]: Pls add a timeout.

**Review by ErikEJ — CHANGES_REQUESTED** [review]

File: src/GUI/EFCorePowerTools/ViewModels/ObjectTreeViewModel.cs (hunk 1 above)
Comment (sm15455) [reply src/GUI/EFCorePowerTools/ViewModels/ObjectTreeViewModel.cs:57]: What do you mean by "timeout"? Something like: "if the search is not over within 3 seconds abort it?" or something else?

Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/ViewModels/ObjectTreeViewModel.cs:57]: Yes, it is a security issue to not add a timeout, say 5 seconds (via RegExOptions)

Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/ViewModels/ObjectTreeViewModel.cs:57]: https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices#use-time-out-values

Comment (sm15455) [reply src/GUI/EFCorePowerTools/ViewModels/ObjectTreeViewModel.cs:57]: This days isn't useless, I've learnt something new :D

File: src/GUI/EFCorePowerTools/Dialogs/PickTablesDialog.xaml [hunk 2]
```diff
@@ -107,7 +107,59 @@
                      HorizontalAlignment="Stretch"
                      VerticalContentAlignment="Center"
                      Text="{Binding SearchText, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}" 
-                     Grid.Column="2" />
+                     Grid.Column="2" 
+                     x:Name="Search">
+                <!--<TextBox.Triggers>
```
Comment (ErikEJ) [inline src/GUI/EFCorePowerTools/Dialogs/PickTablesDialog.xaml:112]: Remove this?

File: src/GUI/EFCorePowerTools/ViewModels/ObjectTreeViewModel.cs [hunk 3]
```diff
@@ -41,11 +42,27 @@ public class ObjectTreeViewModel : ViewModelBase, IObjectTreeViewModel
...
+        public void Search(string searchText, SearchMode searchMode)
         {
+            var regex = new Regex(searchText, RegexOptions.None, new TimeSpan(0,0,3));
+
             foreach (var obj in _objects)
             {
-                obj.IsVisible = string.IsNullOrWhiteSpace(searchText) || obj.DisplayName.ToUpper().Contains(searchText.ToUpper());
+                if (searchMode == SearchMode.Text)
+                {
+                    obj.IsVisible = string.IsNullOrWhiteSpace(searchText) || obj.DisplayName.ToUpper().Contains(searchText.ToUpper());
```
Comment (ErikEJ) [inline src/GUI/EFCorePowerTools/ViewModels/ObjectTreeViewModel.cs:53]: Use:
`obj.DisplayName.IndexOf(searchText, StringComparison.OrdinalIgnoreCase) >= 0;`

**Review by ErikEJ — CHANGES_REQUESTED** [review]

File: src/GUI/.editorconfig [hunk 4]
```diff
@@ -0,0 +1,152 @@
+# Rules in this file were initially inferred by Visual Studio IntelliCode from the C:\Work\SMProjects\contributions\EFCorePowerTools\src\GUI codebase based on best match to current usage at 30/01/2021
```
Comment (ErikEJ) [inline src/GUI/.editorconfig:1]: Please remove :smile:

File: src/GUI/EFCorePowerTools/Dialogs/PickTablesDialog.xaml [hunk 5]
```diff
@@ -107,7 +107,44 @@
...
+            <Popup x:Name="ContextPopup" 
+                   PlacementTarget="{Binding ElementName=MainButton}" 
+                   Placement="Left" 
+                   VerticalOffset="{Binding ActualHeight, ElementName=MainButton}" 
+                   HorizontalOffset="{Binding ActualWidth, ElementName=MainButton}" 
+                   StaysOpen="False">
+                <Border BorderThickness="1" BorderBrush="LightGray" Background="{StaticResource DialogWindowBackgroundColor}">
+                    <StackPanel>
+                        <RadioButton Margin="8"
+                                     IsChecked="{Binding SearchMode, Converter={StaticResource EnumToBoolConverter}, ConverterParameter={x:Static contractvm:SearchMode.Text}}">Use text search</RadioButton>
```
Comment (ErikEJ) [inline src/GUI/EFCorePowerTools/Dialogs/PickTablesDialog.xaml:142]: espression => expression

**Review by ErikEJ — CHANGES_REQUESTED** [review]

**Comment by ErikEJ** [comment]
Thanks!

**Comment by sm15455** [comment]
Always a pleasure ;)

**Pull request merged**